spark/v3.2/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/RewriteUpdateTable.scala

```diff
@@ -0,0 +1,157 @@
+    case u @ UpdateIcebergTable(aliasedTable, assignments, cond, None) if u.resolved =>
+      EliminateSubqueryAliases(aliasedTable) match {
+        case r @ DataSourceV2Relation(tbl: SupportsRowLevelOperations, _, _, _, _) =>
+          val operation = buildRowLevelOperation(tbl, UPDATE)
+          val table = RowLevelOperationTable(tbl, operation)
+          val rewritePlan = operation match {
+            case _: SupportsDelta =>
+              throw new AnalysisException("Delta updates are currently not supported")
+            case _ if cond.exists(SubqueryExpression.hasSubquery) =>
+              buildReplaceDataWithUnionPlan(r, table, assignments, cond)
+            case _ =>
+              buildReplaceDataPlan(r, table, assignments, cond)
+          }
+          UpdateIcebergTable(r, assignments, cond, Some(rewritePlan))
+
+        case p =>
+          throw new AnalysisException(s"$p is not an Iceberg table")
+      }
+  }
+
+  // build a rewrite plan for sources that support replacing groups of data (e.g. files, partitions)
+  // if the condition does NOT contain a subquery
+  private def buildReplaceDataPlan(
+      relation: DataSourceV2Relation,
+      table: RowLevelOperationTable,
+      assignments: Seq[Assignment],
+      cond: Option[Expression]): ReplaceData = {
+
+    // resolve all needed attrs (e.g. metadata attrs for grouping data on write)
+    val metadataAttrs = resolveRequiredMetadataAttrs(relation, table.operation)
+
+    // construct a read relation and include all required metadata columns
+    val readAttrs = dedupAttrs(relation.output ++ metadataAttrs)
+    val readRelation = relation.copy(table = table, output = readAttrs)
+
+    // build a plan with updated rows
+    val updateCond = cond.getOrElse(Literal.TrueLiteral)
+    val allRowsPlan = buildUpdateProjection(readRelation, assignments, updateCond)
+
+    // build a plan to replace read groups in the table
+    val writeRelation = relation.copy(table = table)
+    ReplaceData(writeRelation, allRowsPlan, relation)
+  }
+
+  // build a rewrite plan for sources that support replacing groups of data (e.g. files, partitions)
+  // if the condition contains a subquery
+  private def buildReplaceDataWithUnionPlan(
+      relation: DataSourceV2Relation,
+      table: RowLevelOperationTable,
+      assignments: Seq[Assignment],
+      cond: Option[Expression]): ReplaceData = {
+
+    // resolve all needed attrs (e.g. metadata attrs for grouping data on write)
+    val metadataAttrs = resolveRequiredMetadataAttrs(relation, table.operation)
+
+    // construct a read relation and include all required metadata columns
+    // the same read relation will be used to read records that must be updated and be copied over
+    // DeduplicateRelations will take care of duplicated attr IDs
+    val readAttrs = dedupAttrs(relation.output ++ metadataAttrs)
+    val readRelation = relation.copy(table = table, output = readAttrs)
```

Comment:
This part requires a bit of time to understand. The comment above is also important.

In short, we are using the same relation on both sides of the union. Then `DeduplicateRelations` clones one of them and assigns fresh expression IDs so that attributes on both sides are different. During job planning, we plan the job only once and then replace both `DataSourceV2Relation` with two `DataSourceV2ScanRelation`. The scan relations are different objects but they share the SAME `Scan` instance. During dynamic filtering, we assign each scan relation a dynamic filter. The dynamic filter subquery will be executed only once and then shared by two dynamic filters at runtime. Luckily, Spark reuses subqueries with and without AQE. The reason why we have two dynamic filters is because we don't know which side of the union plan will be executed first.